curriculum/challenges/english/25-front-end-development/lab-hash-table/67ed03ac474c48692f41749e.md

```diff
@@ -0,0 +1,294 @@
+
+    ```md
+    {
+        300: {
+            "fcc": "coding",
+            "cfc": "chemical"
+        }
+    }
+    ```
+
+# --hints--
+
+You should define the `HashTable` class.
+
+```js
+({ 
+    test: () => assert(runPython(`
+    _Node(_code).has_class("HashTable")
+    `)) 
+})
+```
+
+The `HashTable` should initialize a `collection` property as an empty dictionary within the `__init__` method.
+
+```js
+({
+    test: () => assert(runPython(`
+      _Node(_code).find_class("HashTable").find_function("__init__").has_stmt("self.collection={}")
+    `))
+})
+```
+
+The `HashTable` class should have a `hash` method.
+
+```js
+({test: () => {assert(runPython(`_Node(_code).find_class('HashTable').has_function('hash')`))}})
+```
+
+The `hash` method should take `string` as a parameter.
+
+```js
+({ test: () => assert(runPython(`_Node(_code).find_class("HashTable").find_function("hash").has_args("self,string")`)) })
+```
+
+The `hash` method should return the sum of the Unicode values of each character in the string.
+
+```js
+({ 
+    test: () => runPython(`
+        ht = HashTable()
+        hash_result = ht.hash("fcc")
+        assert hash_result == 300
+
+        # prevent hardcoding
+        assert ht.hash("golf") == 424
+        assert ht.hash("read") == 412
+`) })
+```
+
+The `HashTable` class should have an `add` method.
+
+```js
+({test: () => {assert(runPython(`_Node(_code).find_class('HashTable').has_function('add')`))}})
+```
+
+The `add` method should take `key` and `value` as parameters.
+
+```js
+({ test: () => assert(runPython(`_Node(_code).find_class("HashTable").find_function("add").has_args("self,key,val")`)) })
+```
+
+The `HashTable` class should have a `remove` method.
+
+```js
+({test: () => {assert(runPython(`_Node(_code).find_class('HashTable').has_function('remove')`))}})
+```
+
+The `remove` method should take `key` as a parameter.
+
+```js
+({ test: () => assert(runPython(`_Node(_code).find_class("HashTable").find_function("remove").has_args("self,key")`)) })
+```
+
+The `HashTable` class should have a `lookup` method.
+
+```js
+({test: () => {assert(runPython(`_Node(_code).find_class('HashTable').has_function('lookup')`))}})
+```
+
+The `lookup` method should take `key` as the parameter.
+
+```js
+({ test: () => assert(runPython(`_Node(_code).find_class("HashTable").find_function("lookup").has_args("self,key")`)) })
+```
+
+`HashTable.hash("golf")` should return `424`.
+
+```js
+({ 
+    test: () => runPython(`
+        ht = HashTable()
+        hash_result = ht.hash("golf")
+        assert hash_result == 424
+
+        # prevent hardcoding
+        assert ht.hash("dear") == 412
+        assert ht.hash("cat") == 312
+    `) 
+})
+```
+
+`HashTable.add("golf", "sport")` should add the key-value pair to the collection at index `424`.
+
+```js
+
+({ 
+    test: () => runPython(`
+    ht = HashTable()
+    ht.add("golf", "sport")
+    expected_value = {424: {'golf': 'sport'}}
+
+    assert ht.collection == expected_value
```

Comment:
This is never satisfied because my dictionary will never be that dictionary. I haven't checked the remaining tests yet.

Reply:
```py
d1 = {1: {2: "two"}}
d2 = {1: {2: "three"}}
def deepEqualDict(d1, d2):
    for (k1,v1), (k2,v2) in zip(d1.items(), d2.items()):
        if k1 != k2:
            return False
        if type(v1) == dict and type(v2) == dict:
            if not deepEqualDict(v1, v2):
                return False
        elif v1 != v2:
            return False
    
    return True
        
        
deepEqualDict(d1, d2)
```

a function like this would be needed for testing if two dictionaries are the same

Reply:
You're right, my bad :facepalm: I was too sure about that because the test was failing but everything seemed fine. I had a typo...